pkg/approver/approver.go

```diff
@@ -0,0 +1,188 @@
+package approver
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	configv1 "github.com/openshift/api/config/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog"
+
+	certificatesv1 "k8s.io/api/certificates/v1beta1"
+
+	"github.com/openshift/library-go/pkg/config/clusteroperator/v1helpers"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	clusterOperatorName           = "machine-approver"
+	clusterOperatorNamespace      = "openshift-cluster-machine-approver"
+	unknownVersionValue           = "unknown"
+	operatorVersionKey            = "operator"
+	reasonAsExpected              = "AsExpected"
+	releaseVersionEnvVariableName = "RELEASE_VERSION"
+)
+
+var relatedObjects = []configv1.ObjectReference{
+	{
+		Group:    "",
+		Resource: "namespaces",
+		Name:     clusterOperatorNamespace,
+	},
+	{
+		Group:    "certificates.k8s.io",
+		Resource: "certificatesigningrequests",
+		Name:     "",
+	},
+}
+
+type StatusController struct {
+	client.Client
+}
+
+func (c *StatusController) SetupWithManager(mgr ctrl.Manager, options controller.Options) error {
+	return setupWithManager(mgr, options).Complete(c)
+}
+
+func setupWithManager(mgr ctrl.Manager, options controller.Options) *builder.Builder {
```

Comment:
Why not pass the `reconcile.Reconciler` into this and return an error? Feels a bit weird to me to be exposing the builder here. If we do it this way around you don't need to remember to call `Complete` on it when you're setting up new tests

I had expected `SetupWithManager` to look more like:

func (c *StatusController) SetupWithManager(mgr ctrl.Manager, options controller.Options) error {
	return setupWithManager(mgr, options, c)
}